src/descriptor/mod.rs

```diff
@@ -225,6 +231,34 @@ impl DescriptorPublicKey {
             ))
         }
     }
+
+    /// Derives a new key using the path if self is a wildcard xpub. Otehrwise returns a copy of
+    /// self.
+    ///
+    /// Panics if derivation path contains a hardened child number
+    pub fn derive(&self, path: &[bip32::ChildNumber]) -> DescriptorPublicKey {
```

Comment:
There is too much cloning going on here, should we consume `self` and let the caller decide whether to clone or not?

Reply:
Yeah, definitely seems a better API.

Reply:
I'd generally agree, but the primary use case is to do multiple derivations e.g. from a `DerivationPathIterator`, not a single derivation. So you'd nearly always clone anyway. Maybe a key type with a lifetime that only borrows the additional derivation path would be a solution? I just tried to keep it simple for now by not over-optimizing it.

Reply:
Came up with
```rust
diff --git a/src/descriptor/mod.rs b/src/descriptor/mod.rs
index 49ecae2..00a780c 100644
--- a/src/descriptor/mod.rs
+++ b/src/descriptor/mod.rs
@@ -101,7 +101,7 @@ impl fmt::Display for DescriptorKeyParseError {
 impl fmt::Display for DescriptorPublicKey {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match *self {
-            DescriptorPublicKey::PukKey(ref pk) => pk.fmt(f),
+            DescriptorPublicKey::PubKey(ref pk) => pk.fmt(f),
             DescriptorPublicKey::XPub(ref xpub) => {
                 if let Some((ref master_id, ref master_deriv)) = xpub.origin {
                     fmt::Formatter::write_str(f, "[")?;
@@ -232,21 +232,22 @@ impl DescriptorPublicKey {
         }
     }
 
-    /// Derives a new key using the path if self is a wildcard xpub. Otehrwise returns a copy of
+    /// Derives a new key using the path if self is a wildcard xpub. Otherwise returns a copy of
     /// self.
     ///
     /// Panics if derivation path contains a hardened child number
-    pub fn derive(&self, path: &[bip32::ChildNumber]) -> DescriptorPublicKey {
+    pub fn derive(self, path: &[bip32::ChildNumber]) -> DescriptorPublicKey {
         assert!(path.into_iter().all(|c| c.is_normal()));
 
-        match *self {
-            DescriptorPublicKey::PukKey(ref pk) => DescriptorPublicKey::PukKey(*pk),
-            DescriptorPublicKey::XPub(ref xpub) => {
+        match self {
+            DescriptorPublicKey::PubKey(_) => self,
+            DescriptorPublicKey::XPub(xpub) => {
                 if xpub.is_wildcard {
                     DescriptorPublicKey::XPub(DescriptorXPub {
-                        origin: xpub.origin.clone(),
-                        xpub: xpub.xpub.clone(),
-                        derivation_path: (&xpub.derivation_path)
+                        origin: xpub.origin,
+                        xpub: xpub.xpub,
+                        derivation_path: xpub
+                            .derivation_path
                             .into_iter()
                             .chain(path.iter())
                             .cloned()
@@ -254,7 +255,7 @@ impl DescriptorPublicKey {
                         is_wildcard: false,
                     })
                 } else {
-                    self.clone()
+                    DescriptorPublicKey::XPub(xpub)
                 }
             }
         }
@@ -272,7 +273,7 @@ impl MiniscriptKey for DescriptorPublicKey {
 impl ToPublicKey for DescriptorPublicKey {
     fn to_public_key(&self) -> bitcoin::PublicKey {
         match *self {
-            DescriptorPublicKey::PukKey(ref pk) => *pk,
+            DescriptorPublicKey::PubKey(ref pk) => *pk,
             DescriptorPublicKey::XPub(ref xpub) => {
                 let ctx = secp256k1::Secp256k1::verification_only();
                 xpub.xpub
@@ -651,7 +652,7 @@ impl Descriptor<DescriptorPublicKey> {
     /// Derives all wildcard keys in the descriptor using the supplied `path`
     pub fn derive(&self, path: &[bip32::ChildNumber]) -> Descriptor<DescriptorPublicKey> {
         self.translate_pk(
-            |pk| Result::Ok::<DescriptorPublicKey, ()>(pk.derive(path)),
+            |pk| Result::Ok::<DescriptorPublicKey, ()>(pk.clone().derive(path)),
             |pkh| Result::Ok::<hash160::Hash, ()>(*pkh),
         )
         .expect("Translation fn can't fail.")
```
For bf13c51df93399f0115c6d72e1f4cecf00219701, applying it for now but we can still discuss it.

Reply:
> Otherwise returns ~a copy of~ self.

Reply:
Thanks, corrected

Reply:
Was a good idea finally, this allows to remove the `clone()` in `derive`.